src/Chainweb/Pact/Service/PactQueue.hs

```diff
@@ -51,11 +99,102 @@ addRequest q msg =  atomically $ writeTBQueue priority msg
+getNextRequest q = do
+  (T2 req entranceTime) <- atomically $
     tryReadTBQueueOrRetry (_pactQueueValidateBlock q)
     <|> tryReadTBQueueOrRetry (_pactQueueNewBlock q)
     <|> tryReadTBQueueOrRetry (_pactQueueOtherMsg q)
+  exitTime <- getCurrentTimeIntegral
+  let requestTime = exitTime `diff` entranceTime
+      stats = case req of
+        ValidateBlockMsg {} -> _pactQueuePactQueueValidateBlockMsgStats q
+        NewBlockMsg {} -> _pactQueuePactQueueNewBlockMsgStats q
+        _ -> _pactQueuePactQueueOtherMsgStats q
+  updatePactQueueStats stats requestTime
+  return req
   where
     tryReadTBQueueOrRetry = tryReadTBQueue >=> \case
       Nothing -> retry
       Just msg -> return msg
+
+
+data PactQueueStats = PactQueueStats
+  {
+    _pactQueueStatsQueueName :: !T.Text
+  , _pactQueueStatsCounters  :: !(IORef PactQueueCounters)
+  }
+
+data PactQueueCounters = PactQueueCounters
+  {
+    _pactQueueCountersCount :: {-# UNPACK #-} !Int
+  , _pactQueueCountersSum   :: {-# UNPACK #-} !Micros
+  , _pactQueueCountersMin   :: {-# UNPACK #-} !Micros
+  , _pactQueueCountersMax   :: {-# UNPACK #-} !Micros
+  } deriving (Show, Generic)
+    deriving anyclass NFData
+
+
+instance ToJSON PactQueueCounters where
+  toJSON (PactQueueCounters {..}) =
+    object
+      [
+        "count" .= _pactQueueCountersCount
+      , "sum"   .= _pactQueueCountersSum
+      , "min"   .= _pactQueueCountersMin
+      , "max"   .= _pactQueueCountersMax
+      , "avg"   .= avg
+      ]
+    where
+      avg :: Maybe Double
+      avg = if _pactQueueCountersCount == 0 then Nothing
+        else Just $ fromIntegral _pactQueueCountersSum / fromIntegral _pactQueueCountersCount
+
+updatePactQueueStats :: PactQueueStats -> TimeSpan Micros -> IO ()
+updatePactQueueStats stats (timeSpanToMicros -> timespan) = do
+    atomicModifyIORef' (_pactQueueStatsCounters stats) $ \ctrs ->
+      (PactQueueCounters
+      {
+          _pactQueueCountersCount = _pactQueueCountersCount ctrs + 1
+        , _pactQueueCountersSum = _pactQueueCountersSum ctrs + timespan
+        , _pactQueueCountersMin = _pactQueueCountersMin ctrs `min` timespan
+        , _pactQueueCountersMax = _pactQueueCountersMax ctrs `max` timespan
+      }
+      , ())
+
+resetPactQueueStats :: PactQueue -> IO ()
+resetPactQueueStats q = do
+  resetPactQueueStats' (_pactQueuePactQueueValidateBlockMsgStats q)
+  resetPactQueueStats' (_pactQueuePactQueueNewBlockMsgStats q)
+  resetPactQueueStats' (_pactQueuePactQueueOtherMsgStats q)
+
+resetPactQueueStats' :: PactQueueStats -> IO ()
+resetPactQueueStats' stats = atomicWriteIORef (_pactQueueStatsCounters stats) initPactQueueCounters
+
+getPactQueueStats :: PactQueue -> IO (Value, Value, Value)
+getPactQueueStats = getPactQueueStats' >=> \case
+    (vstats',nbstats',ostats') -> do
+      Time timestamp  <- getCurrentTimeIntegral @Micros
+      let
+        vstats  = withObject' (toJSON vstats') ("msg_type" .= ("validate" :: T.Text) <> "timestamp" .= timestamp)
```

Comment:
Why not using an `ToJSON` instance for `PactQueueStats`?

Also, I am not sure whether the timestamp is needed, since each log message receives a timestamp when it is created.